homeassistant/components/hko/__init__.py

```diff
@@ -0,0 +1,260 @@
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+
+from async_timeout import timeout
+from hko import HKO, LOCATIONS, HKOError
+
+from homeassistant.components.weather import (
+    ATTR_CONDITION_CLEAR_NIGHT,
+    ATTR_CONDITION_CLOUDY,
+    ATTR_CONDITION_FOG,
+    ATTR_CONDITION_HAIL,
+    ATTR_CONDITION_LIGHTNING_RAINY,
+    ATTR_CONDITION_PARTLYCLOUDY,
+    ATTR_CONDITION_POURING,
+    ATTR_CONDITION_RAINY,
+    ATTR_CONDITION_SNOWY,
+    ATTR_CONDITION_SNOWY_RAINY,
+    ATTR_CONDITION_SUNNY,
+    ATTR_CONDITION_WINDY,
+    ATTR_CONDITION_WINDY_VARIANT,
+    ATTR_FORECAST_CONDITION,
+    ATTR_FORECAST_TEMP,
+    ATTR_FORECAST_TEMP_LOW,
+    ATTR_FORECAST_TIME,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_LOCATION, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    API_CURRENT,
+    API_DATA,
+    API_FORECAST,
+    API_FORECAST_DATE,
+    API_FORECAST_ICON,
+    API_FORECAST_MAX_TEMP,
+    API_FORECAST_MIN_TEMP,
+    API_FORECAST_WEATHER,
+    API_HUMIDITY,
+    API_PLACE,
+    API_TEMPERATURE,
+    API_VALUE,
+    API_WEATHER_FORECAST,
+    DEFAULT_DISTRICT,
+    DOMAIN,
+    KEY_DISTRICT,
+    KEY_LOCATION,
+    WEATHER_INFO_AT_TIMES_AT_FIRST,
+    WEATHER_INFO_CLOUD,
+    WEATHER_INFO_FINE,
+    WEATHER_INFO_FOG,
+    WEATHER_INFO_HEAVY,
+    WEATHER_INFO_INTERVAL,
+    WEATHER_INFO_ISOLATED,
+    WEATHER_INFO_MIST,
+    WEATHER_INFO_OVERCAST,
+    WEATHER_INFO_PERIOD,
+    WEATHER_INFO_RAIN,
+    WEATHER_INFO_SHOWER,
+    WEATHER_INFO_SNOW,
+    WEATHER_INFO_SUNNY,
+    WEATHER_INFO_THUNDERSTORM,
+    WEATHER_INFO_WIND,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[Platform] = [Platform.WEATHER]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Hong Kong Observatory from a config entry."""
+
+    location = entry.data[CONF_LOCATION]
+    district = next(
+        (item for item in LOCATIONS if item[KEY_LOCATION] == location),
+        {KEY_DISTRICT: DEFAULT_DISTRICT},
+    )[KEY_DISTRICT]
+    websession = async_get_clientsession(hass)
+
+    coordinator = HKOUpdateCoordinator(hass, websession, district, location)
+    await coordinator.async_config_entry_first_refresh()
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class HKOUpdateCoordinator(DataUpdateCoordinator):
```

Comment:
To my understanding the Coordinator should be in its own coordinator.py file

Reply:
Fixed.